instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go

```diff
@@ -173,14 +284,27 @@ func handleRPC(ctx context.Context, rs stats.RPCStats) {
 			),
 		)
 	case *stats.End:
+		var rpcStatusAttr attribute.KeyValue
+		metricAttrs := make([]attribute.KeyValue, 0, len(gctx.metricAttrs)+1)
+		metricAttrs = append(metricAttrs, gctx.metricAttrs...)
+
 		if rs.Error != nil {
 			s, _ := status.FromError(rs.Error)
 			span.SetStatus(codes.Error, s.Message())
-			span.SetAttributes(statusCodeAttr(s.Code()))
+			rpcStatusAttr = semconv.RPCGRPCStatusCodeKey.Int(int(s.Code()))
 		} else {
-			span.SetAttributes(statusCodeAttr(grpc_codes.OK))
+			rpcStatusAttr = semconv.RPCGRPCStatusCodeKey.Int(int(grpc_codes.OK))
 		}
+		span.SetAttributes(rpcStatusAttr)
 		span.End()
+
+		metricAttrs = append(metricAttrs, rpcStatusAttr)
+		r.rpcDuration.Record(context.TODO(), int64(rs.EndTime.Sub(rs.BeginTime)), metric.WithAttributes(metricAttrs...))
```

Comment:
Shouldn't we use the `ctx` passed to `handleRPC`?

Reply:
I have tried to use `ctx`, but it won't work, I'm still digging the reason.

Reply:
I found the cause. 

The reason is that, when we add instrument in `HandleRPC` function, when the `rs stats.RPCStats` is type of `*stats.End` and `*stats.OutTrailer`, the context we get is already cancelled by in grpc-go sdk. 

https://github.com/grpc/grpc-go/blob/6e9c88b0acf15de2256c3d3a257869e0669572d9/internal/transport/http2_server.go#L1277

```go
func (t *http2Server) finishStream(s *Stream, rst bool, rstCode http2.ErrCode, hdr *headerFrame, eosReceived bool) {
	// In case stream sending and receiving are invoked in separate
	// goroutines (e.g., bi-directional streaming), cancel needs to be
	// called to interrupt the potential blocking on other goroutines.
	s.cancel()

	oldState := s.swapState(streamDone)
```

And the metric sdk will check if ctx.Err() is nil when record the measurement
https://github.com/open-telemetry/opentelemetry-go/blob/c047088605b454f172765621d53107f80b1e6417/sdk/metric/instrument.go#L205
```
func (i *int64Inst) aggregate(ctx context.Context, val int64, s attribute.Set) { // nolint:revive  // okay to shadow pkg with method.
	if err := ctx.Err(); err != nil {
		return
	}
	for _, in := range i.measures {
		in(ctx, val, s)
	}
}
```

So the aggregate function will fail, the measurement will also be dropped.

So we should 1. use new context object. 2. add the measurement before `*stats.End` and `*stats.OutTrailer`

But some metric can only be measured when it comes to `*stats.End` and `*stats.OutTrailer`, for example duration, requests per rpc and responses per rpc. (Since in streaming model, we can only know the final count of requests and response when it comes to `*stats.End`). So I vote for the solution 1.

Reply:
> And the metric sdk will check if ctx.Err() is nil when record the measurement
https://github.com/open-telemetry/opentelemetry-go/blob/c047088605b454f172765621d53107f80b1e6417/sdk/metric/instrument.go#L205

I am not convinced that this what SDK should be doing. @MrAlias WDYT?

Reply:
I created https://github.com/open-telemetry/opentelemetry-go/pull/4671

Reply:
This would be a better solution: https://github.com/open-telemetry/opentelemetry-go/pull/4671#issuecomment-1783534940